crates/hcl-edit/src/parser/expr.rs

```diff
@@ -54,53 +54,47 @@ fn expr_inner<'i, 's>(
-                        continue;
-                    }
-                    // Conditional.
-                    [b'?', _] => {
-                        state.borrow_mut().on_ws(suffix);
-                        return conditional(state).parse_next(input);
-                    }
-                    // Binary operation.
-                    //
-                    // Note: matching a single `=` is ambiguous as it could also be an object
-                    // key-value separator, so we'll need to match on `==`.
-                    b"=="
-                    | [b'!' | b'<' | b'>' | b'+' | b'-' | b'*' | b'/' | b'%' | b'&' | b'|', _] => {
-                        state.borrow_mut().on_ws(suffix);
-                        return binary_op(state).parse_next(input);
-                    }
-                    // None of the above matched.
-                    _ => {
-                        input.reset(checkpoint);
-                        return Ok(());
-                    }
+            // Peek the next two bytes to identify the following operation, if any.
+            match peek(take::<_, _, ContextError>(2usize)).parse_next(input) {
+                // The sequence `..` might introduce a `...` operator within a for object expr
+                // or after the last argument of a function call, do not mistakenly parse it as
+                // a `.` traversal operator.
+                //
+                // `//` and `/*` are comment starts. Do not mistakenly parse a `/` as binary
+                // division operator.
+                Ok(b"//" | b"/*" | b"..") => {
+                    input.reset(checkpoint);
+                    return Ok(());
+                }
+                // Traversal operator.
+                //
+                // Note: after the traversal is consumed, the loop is entered again to consume
+                // a potentially following conditional or binary operation.
+                Ok([b'.' | b'[', _]) => {
+                    state.borrow_mut().on_ws(suffix);
+                    traversal(state).parse_next(input)?;
+                    continue;
+                }
+                // Conditional.
+                Ok([b'?', _]) => {
+                    state.borrow_mut().on_ws(suffix);
+                    return conditional(state).parse_next(input);
+                }
+                // Binary operation.
+                //
+                // Note: matching a single `=` is ambiguous as it could also be an object
+                // key-value separator, so we'll need to match on `==`.
+                Ok(b"==")
+                | Ok([b'!' | b'<' | b'>' | b'+' | b'-' | b'*' | b'/' | b'%' | b'&' | b'|', _]) => {
+                    state.borrow_mut().on_ws(suffix);
+                    return binary_op(state).parse_next(input);
+                }
+                // None of the above matched or we hit the end of input.
+                _ => {
+                    input.reset(checkpoint);
```

Comment:
Would wrapping the `peek` in `opt` work?  I tended to use `opt` to avoid manually calling `reset` in toml_edit

Reply:
I was trying the same, but it does not work since there's some ambiguity in the grammar (e.g. the first case in the match which has overlap with the last one) where we need to rewind the parser to the point before the whitespace we parsed before the peek. E.g. something like `match opt((sp.span(), peek(take(2usize)))).parse_next(input)? { ... }` will still require a manual `input.reset(checkpoint)` in these match branches.

Instead of `take(2usize)` I could of course use a parser that would exclude these ambiguous cases, but it would probably involve `alt` and will become more complex than it already is due to the overlap in the grammar. Performance will probably also suffer quite a bit since this function is in the parser's hot path.

I'm totally fine with `input.reset(checkpoint)` here, it's much easier to reason about than how things were pre-0.5.0 😃.

Reply:
I'll probably still switch to `match opt((sp.span(), peek(take(2usize)))).parse_next(input)? { ... }` as it makes the function a tiny bit shorter, avoids the turbofish on `take` and doesn't seem to negatively affect performance.